File: src/medium/LongestPalindrome.java
package medium;

public class LongestPalindrome {
  public String longestPalindrome(String s) {
    if (s == null || s.length() < 2) {
      return s;
    }
    int start = 0, end = 0;
    for (int i = 0; i < s.length(); i++) {
      int rs1 = expandAroundCenter(s, i, i);
      int rs2 = expandAroundCenter(s, i, i + 1);
      int rs = Math.max(rs1, rs2);
      if (rs > end - start) {
        end = i + rs / 2;
        start = i - (rs - 1) / 2;
      }
    }
    return s.substring(start, end + 1);
  }

  // return left and right index of the max palindromic string
  private int expandAroundCenter(String s, int l, int r) {
    while (l >= 0 && r < s.length() && s.charAt(l) == s.charAt(r)) {
      l--;
      r++;
    }
    return r - l - 1;
  }
}

File: csharp/LeetCode/Palindrome/LongestPalindrome.cs
namespace LeetCode.Palindrome.LongestPalindrome
{
    using System;

    public class Solution
    {
        public string LongestPalindrome(string s)
        {
            var start = 0;
            var len = 0;

            for (var i = 0; i < s.Length; i++)
            {
                var (l1, l2) = Max(s, i, i);
                var (r1, r2) = Max(s, i, i + 1);
                var max = Math.Max(l2, r2);
                if (max > len)
                {
                    len = max;
                    start = l2 > r2 ? l1 : r1;
                }
            }
            return s.Substring(start, len);
        }


        (int, int) Max(string s, int l, int r)
        {
            while (l >= 0 && r < s.Length && s[l] == s[r])
            {
                l--;
                r++;
            }

            return (l + 1, r - l - 1);
        }
    }
}
